index.md

```diff
@@ -1,16 +1,128 @@
+
+The first method is closer to a known quantity.  However, if a payload is not already packaged as a container, then it ends up being a variation on the second method.  The most worrying thing about this approach is that it makes accomodating new payloads harder, because significant portions of the non-payload pieces need to be rewritten for each payload.
+
+The second method means, at least theoretically, that we could keep one lab-building framework, and drop in new payloads by simply updating the single build stage that implement the installation (and runtime launch framework) of that payload.  The problem there is that there is no universal method of doing so.  However, it seems difficult to imagine that one of the following five techniques is not applicable.
+
+1. The payload is pip-installable: there's a `requirements.txt` file, the installer points `pip` at it, and the payload is installed.
+2. The payload is conda-installable: there's an environment `YAML` file, the installer points `conda` at it, and the payload is installed.
+3. The payload is supplied as a Dockerfile that constructs a container image.  The appropriate pieces of the Dockerfile can be moved into the nublado-launchable image construction (either as `COPY` and `RUN` statements, or as a shell script which does the same thing), and the payload is installed.
+4. The payload is only available as a Docker image, without build instructions.  This is alarming and certainly suggests the first method would be a better idea, but assuming that the payload is limited somehow to a particular place in the filesystem (for instance, the DM pipelines stack lives almost entirely under `/opt/lsst/software/stack`), a multi-stage Docker build could transplant that directory hierarchy into the target container.
+5. The payload is delivered as an installation script.  The DM pipelines stack can also be installed this way with [lsstinstall][2] and this may be the most common way for thorny environments to be delivered.  For anything of sufficent size, that has made the (bad) choice to not use a standard build/installation system, this will probably already exist.  See, for instance, the huge proliferation in products that expect you to `curl https://some-url | sudo /bin/bash`.  Less grotesquely, maybe it's something you can install with `./configure; make; make install`.
```

Comment:
Here too, I think this needs to come with the caveat that the installation script cannot assume a specific version of Linux and needs to restrain its actions to a specific tree on the file system.